### PR TITLE
fix(tooltip): resolve `overflow: hidden` issue

### PR DESCRIPTION
**Pull request by cbacon91**
## Motivation
Fix issue in tooltip where `overflow: hidden` makes it impossible for the tooltip to be seen in all of its glory.

## PR upkeep checklist

- [x] Change log entry
- [x] Label(s)
- [x] Assignee(s)
- [x] Deployment URL: https://mavenlink.github.io/design-system/webinf-portal-just-the-tip/
- [x] (When ready for review) Reviewer(s)
- [x] Green Bigmaven CI check
- [x] DevQA on Bigmaven staging (e.g. does the work need to be imported in the internal design system?)

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -11,6 +11,9 @@ module.exports = {
     'import/extensions': ['error', 'always', {
       ignorePackages: true,
     }],
+    'no-mixed-operators': ['error', {
+      allowSamePrecedence: true,
+    }],
```
Comment (cbacon91) [inline .eslintrc.js:16]: Happy to revert, but this lint rule made the following arithmetic throw a lint error:

```
const someValue = a + b - c;
```

You _could_ add parentheses to that statement, but adding parentheses there would change the order of operations. You can add them in such a way that it doesn't matter too, but then you're just adding parens for the sake of making lint happy which makes the code uglier. In the real-world example I had, I had something like

```
const value = (x / 2) + (y / 2) - constant;
```

Adding extra parens to this would make it way harder to read imo:

```
const value = ((x / 2) + (y / 2)) - constant;
```

File: src/components/tooltip/tooltip.jsx [hunk 2]
```diff
@@ -21,56 +16,65 @@ export default function Tooltip({
   text,
   truncate,
 }) {
-  const [visible, setVisible] = useState(false);
+  // size of the triangle in pixels; this should be the same as teh value in the linked stylesheet
+  // TODO is there a way to share the variable so its only in one place
+  const triangleHeight = 8;
```
Comment (cbacon91) [inline src/components/tooltip/tooltip.jsx:21]: @juanca I know you told me there was a way to share variables between css and jsx but I couldn't figure out how to do it.

I thought it was just something like `<div styles={{ '--triangle-height': triangleHeight }} />` but when I did that I never saw the value actually used in my css, I might have done something wrong though.. thoughts?

File: src/components/tooltip/use-tooltip-positioning.test.js [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+          },
+        },
+      }));
+
+      act(() => result.current.show({ target }));
+
+      expect(result.current.position).toEqual({
+        left: sourceRect.x + (sourceRect.width / 2) - (tooltipRect.width / 2),
+        top: sourceRect.y - tooltipRect.height - triangleHeight,
+      });
```
Comment (cbacon91) [inline src/components/tooltip/use-tooltip-positioning.test.js:64]: I didn't like using the actual calculations here, but I was worried about using hardcoded numbers as being unclear to future engineers. I can add them with comments though if that's preferred to the arithmetic

File: src/components/tooltip/tooltip.jsx [hunk 4]
```diff
@@ -21,56 +16,72 @@ export default function Tooltip({
...
 
-  const show = () => setVisible(true);
-  const hide = () => setVisible(false);
+  useLayoutEffect(() => {
+    if (disabled || document.querySelector(`[aria-describedby="${id}"]`)) return;
 
-  const classNames = [className];
+    throw new Error('<Tooltip> was used without an element on the DOM being described by it. ' +
+      'Please add `aria-describedby` to the element this tooltip is being used to describe.');
+  }, []);
```
Comment (cbacon91) [inline src/components/tooltip/tooltip.jsx:31]: @juanca this more or less what you had in mind?

We could probably also make it `return` early for production as well if we don't want to hinder clients. It should fail tests for devs locally so I expect it won't realistically hit production, but I could be wrong on that.

I could also change it from `throw` to `console.error` or something.

File: .eslintrc.js (hunk 1 above)
Comment (juanca) [reply .eslintrc.js:16]: 🚀

**Comment by juanca** [comment]
Generally, positioning absolutely against the document can have unintended and hard to decipher consequences. 

a) Positioning while scrolling
b) Positioning with "distorted" document
c) ... I think there are more but if I don't remember then they probably don't matter in the grand scheme of things

---


Aaaand, here are the results:

a) It scrolls with page. 💯 
b) It miscalculated the tooltip position when the document is scrolled _horizontally_ 

image

This might be fixable though!

File: src/components/tooltip/tooltip.jsx [hunk 5]
```diff
@@ -21,56 +16,72 @@ export default function Tooltip({
   text,
   truncate,
 }) {
-  const [visible, setVisible] = useState(false);
+  // size of the triangle in pixels; this should be the same as teh value in the linked stylesheet
+  // TODO is there a way to share the variable so its only in one place
```
Comment (juanca) [inline src/components/tooltip/tooltip.jsx:20]: Shared across how many places?

CSS <=> CSS: use css variables
JS <=> JS: use exported variables
JS <=> CSS: use css variables, set variable in JS on DOM as `style={{ --le-var: '8px' }}`
JS <=> JS <=> CSS: export all the things!

File: src/components/tooltip/tooltip.jsx (hunk 2 above)
Comment (juanca) [reply src/components/tooltip/tooltip.jsx:21]: Huh.... I can lend my 👀  cuz it should work.

File: src/components/tooltip/tooltip.jsx (hunk 4 above)
Comment (juanca) [reply src/components/tooltip/tooltip.jsx:31]: This is 💯 

> won't realistically hit production

Yeah, it's okay. We should have caught it by deployment time. The ID lookup is constant so it should be blazingly fast. 

> I could also change it from throw to console.error or something.

I think this is fine. Throwing an error is good because we want the code to stop running and make the developer fix it ASAP. At some point, we will implement proper React error boundary handling so it won't be as intrusive -- i.e. blank page.

File: src/components/tooltip/tooltip.jsx (hunk 5 above)
Comment (cbacon91) [reply src/components/tooltip/tooltip.jsx:20]: So it's basically what I mentioned here: https://github.com/mavenlink/design-system/pull/333#discussion_r661669634

JS and CSS are what need to share it - I tried using `style` but ran into issues. I might need your help in it.

**Comment by cbacon91** [comment]
> b) Positioning with "distorted" document

Yeah, I did actually notice that, at one point when I zoomed in using the browser to 125 - 200%, but I just tried it again to take a screenshot and it seems to be working as expected. I'll keep an eye out for it otherwise - any ideas on how to reproduce?

//ss

> b) It miscalculated the tooltip position when the document is scrolled _horizontally_

Ah, crap. Yeah that's super easy to fix! - 8c823e76

File: src/components/tooltip/tooltip.jsx (hunk 2 above)
Comment (cbacon91) [reply src/components/tooltip/tooltip.jsx:21]: Ok, figured it out - I just forgot the units 🤦 

I was using `<div styles={{ '--triangle-height': triangleHeight }} />` when what I should have had was `<div styles={{ '--triangle-height': `${triangleHeight}px` }} />`

File: src/components/help-icon/help-icon.test.jsx [hunk 6]
```diff
@@ -10,6 +10,11 @@ describe('HelpIcon', () => {
     text: 'Help Text',
   };
 
+  // There needs to be something on the DOM that the tooltip is describing, so render a dummy div for each test.
+  beforeEach(() => {
```
Comment (juanca) [inline src/components/help-icon/help-icon.test.jsx:14]: A more standardized pattern for "setting up the env for a given test suite" is the `beforeAll`. And a similar clean up in a `afterAll`. I think it's more of a convention. Normally, it wouldn't work for Mavenlink (tm) tests because we are accustomed to cleaning up the whole env between each test due to global pollution.

File: src/components/tooltip/tooltip.jsx [hunk 7]
```diff
@@ -19,58 +14,71 @@ export default function Tooltip({
...
 
-  switch (direction) {
-    case 'top': classNames.push(styles.top); break;
-    case 'bottom': classNames.push(styles.bottom); break;
-    case 'left': classNames.push(styles.left); break;
-    case 'right': classNames.push(styles.right); break;
-    default: break;
-  }
+  // keep the div so enabling/disabling doesn't affect the dom tree or styles
+  if (disabled) return <div>{children}</div>;
```
Comment (juanca) [inline src/components/tooltip/tooltip.jsx:32]: This is missing `styles.wrapper`. Not sure if that would affect the rendering of `children`.

File: src/components/tooltip/tooltip.test.jsx [hunk 8]
```diff
@@ -18,6 +18,24 @@ describe('tooltip', () => {
     expect(document.body).toMatchSnapshot();
   });
 
+  it('errors if the tooltip is not describing any elements', () => {
+    /* eslint-disable no-console */
+    const originalConsoleError = console.error;
```
Comment (juanca) [inline src/components/tooltip/tooltip.test.jsx:23]: Pro-tip: access the window instead of the global console to bypass the linter! (This is a secret)

```suggestion
    const originalConsoleError = window.console.error;
```

File: src/components/help-icon/help-icon.test.jsx (hunk 6 above)
Comment (cbacon91) [reply src/components/help-icon/help-icon.test.jsx:14]: I think in this case it doesn't work as well. Since `cleanup` is called `afterEach` test, and `beforeAll` only happens once, only the first test will pass. Subsequent tests will fail because this div has been unmounted by `cleanup`.

There's a case to be made that since we're calling `render` twice there's already some weirdness, but this is definitely an edge case which was caused by the `useLayoutEffect` added

image

**Review by juanca — APPROVED** [review]

File: src/components/tooltip/tooltip.test.jsx (hunk 8 above)
Comment (cbacon91) [reply src/components/tooltip/tooltip.test.jsx]: 1d149d67

File: src/components/tooltip/tooltip.jsx (hunk 7 above)
Comment (cbacon91) [reply src/components/tooltip/tooltip.jsx]: b46ecc0c

File: src/components/help-icon/help-icon.test.jsx (hunk 6 above)
Comment (juanca) [reply src/components/help-icon/help-icon.test.jsx:14]: Oh, right, the auto-cleanup. 😓 

I guess the other route is injecting a native DOM node yourself. But then we don't get the auto-cleanup. 

/shrug I'd stick with as-is.
